File: java/src/main/java/io/dure/tree/LevelOrderBottom.java
package io.dure.tree;

import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class LevelOrderBottom {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        Queue<TreeNode> queue = new LinkedList<>();
        List<List<Integer>> res = new LinkedList<>();
        List<Integer> level = new LinkedList<>();
        int current = 0;
        int next = 0;
        if (root != null) {
            queue.offer(root);
            current++;
        }
        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            current--;
            level.add(node.val);
            if (node.left != null) {
                queue.offer(node.left);
                next++;
            }
            if (node.right != null) {
                queue.offer(node.right);
                next++;
            }
            if (current == 0) {
                current = next;
                next = 0;
                res.add(0, level);
                level = new LinkedList<>();
            }
        }
        return res;
    }
}

File: go/tree/levelOrderBottom.go
package tree

func levelOrderBottom(root *TreeNode) [][]int {
	var res [][]int
	var queue []*TreeNode
	if root != nil {
		queue = append(queue, root)
	}
	for len(queue) != 0 {
		level := []int{}
		size := len(queue)
		for i := 0; i < size; i++ {
			node := queue[0]
			queue = queue[1:]
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
			level = append(level, node.Val)
		}
		res = append(res, level)

	}
	for i := 0; i < len(res)/2; i++ {
		res[i], res[len(res)-1-i] = res[len(res)-1-i], res[i]
	}
	return res
}
